File: api/src/util/calculator.ts
class Calculator {
  heightCm: number;
  weightKg: number;

  static readonly efficiencyFactor = 0.8;
  static readonly stepLength = 0.415;

  constructor(heightCm: number, weightKg: number) {
    this.heightCm = heightCm;
    this.weightKg = weightKg;
  }

  // Distância (em metros) a partir do número de passos
  stepsToDistance(steps: number): number {
    let stepLength = this.heightCm * Calculator.stepLength;
    let distanceCm = steps * stepLength;
    let distanceMeters = distanceCm / 100;
    return distanceMeters;
  }

  // Quantidade de passos a partir da distância (em metros)
  distanceToSteps(distance: number): number {
    let stepLength = this.heightCm * Calculator.stepLength;
    let distanceCm = distance * 100;
    let steps = distanceCm / stepLength;
    return Math.round(steps);
  }

  // Calorias gastas de acordo com a quantidade de passos
  stepsToCalories(steps: number): number {
    let distanceMeters = this.stepsToDistance(steps);
    let distanceKm = distanceMeters / 1000;
    let calories = distanceKm * this.weightKg * Calculator.efficiencyFactor;
    return calories;
  }

  // Calorias gastas de acordo com a distância (em metros)
  distanceToCalories(distanceMeters: number): number {
    let steps = this.distanceToSteps(distanceMeters);
    return this.stepsToCalories(steps);
  }

  // Quantidade de passos a partir das calorias gastas
  caloriesToSteps(calories: number): number {
    let distanceKm = calories / (this.weightKg * Calculator.efficiencyFactor);
    let distanceMeters = distanceKm * 1000;
    let steps = this.distanceToSteps(distanceMeters);
    return steps;
  }

  // Distância (em metros) a partir das calorias gastas
  caloriesToDistance(calories: number): number {
    let steps = this.caloriesToSteps(calories);
    return this.stepsToDistance(steps);
  }

  //-------------------- Static methods --------------------

  // Distância (em metros) a partir do número de passos
  static staticStepsToDistance(steps: number, heightCm: number): number {
    return new Calculator(heightCm, 0).stepsToDistance(steps);
  }

  // Quantidade de passos a partir da distância (em metros)
  static staticDistanceToSteps(distance: number, heightCm: number): number {
    return new Calculator(heightCm, 0).distanceToSteps(distance);
  }

  // Calorias gastas de acordo com a quantidade de passos
  static staticStepsToCalories(
    steps: number,
    heightCm: number,
    weightKg: number,
  ): number {
    return new Calculator(heightCm, weightKg).stepsToCalories(steps);
  }

  // Calorias gastas de acordo com a distância (em metros)
  static staticDistanceToCalories(
    distanceMeters: number,
    heightCm: number,
    weightKg: number,
  ): number {
    return new Calculator(heightCm, weightKg).distanceToCalories(
      distanceMeters,
    );
  }

  // Quantidade de passos a partir das calorias gastas
  static staticCaloriesToSteps(
    calories: number,
    heightCm: number,
    weightKg: number,
  ): number {
    return new Calculator(heightCm, weightKg).caloriesToSteps(calories);
  }

  // Distância (em metros) a partir das calorias gastas
  static staticCaloriesToDistance(
    calories: number,
    heightCm: number,
    weightKg: number,
  ): number {
    return new Calculator(heightCm, weightKg).caloriesToDistance(calories);
  }
}

File: app/lib/calculator.dart
class Calculator {
  Calculator({
    required this.heightCm,
    required this.weightKg,
  });

  int heightCm;
  int weightKg;

  static const efficiencyFactor = 0.8;
  static const stepLength = 0.415;

  // Distância (em metros) a partir do número de passos
  double stepsToDistance(int steps) {
    // 1 passo = 0.415 da altura
    double stepLength = heightCm * Calculator.stepLength;

    // Distância em centímetros
    double distanceCm = steps * stepLength;

    // Distância em metros
    double distanceMeters = distanceCm / 100;
    return distanceMeters;
  }

  // Quantidade de passos a partir da distância (em metros)
  int distanceToSteps(double distance) {
    // 1 passo = 0.415 da altura
    final double stepLength = heightCm * Calculator.stepLength;

    // Distância em centímetros
    final double distanceCm = distance * 100;

    // Número de passos
    final double steps = distanceCm / stepLength;

    return steps.round();
  }

  // Calorias gastas de acordo com a quantidade de passos
  double stepsToCalories(int steps) {
    final double distanceMeters = stepsToDistance(steps);
    final double distanceKm = distanceMeters / 1000;

    final double calories = distanceKm * weightKg * Calculator.efficiencyFactor;

    return calories;
  }

  // Calorias gastas de acordo com a distância (em metros)
  double distanceToCalories(double distanceMeters) {
    int steps = distanceToSteps(distanceMeters);
    return stepsToCalories(steps);
  }

  // Quantidade de passos a partir das calorias gastas
  int caloriesToSteps(double calories) {
    final double distanceKm =
        calories / (weightKg * Calculator.efficiencyFactor);

    final double distanceMeters = distanceKm * 1000;

    int steps = distanceToSteps(distanceMeters);

    return steps;
  }

  // Distância (em metros) a partir das calorias gastas
  double caloriesToDistance(double calories) {
    int steps = caloriesToSteps(calories);
    return stepsToDistance(steps);
  }

  //--------------------------------------------------

  // Distância (em metros) a partir do número de passos e altura (em centímetros) da pessoa
  static double staticStepsToDistance(int steps, int heightCm) {
    return Calculator(heightCm: heightCm, weightKg: 0).stepsToDistance(steps);
  }

  // Quantidade de passos a partir da distância (em metros)
  static int staticDistanceToSteps(double distance, int heightCm) {
    return Calculator(heightCm: heightCm, weightKg: 0)
        .distanceToSteps(distance);
  }

  // Calorias gastas de acordo com a quantidade de passos
  static double staticStepsToCalories(int steps, int heightCm, int weightKg) {
    return Calculator(heightCm: heightCm, weightKg: weightKg)
        .stepsToCalories(steps);
  }

  // Calorias gastas de acordo com a distância (em metros)
  static double staticDistanceToCalories(
      double distanceMeters, int heightCm, int weightKg) {
    return Calculator(heightCm: heightCm, weightKg: weightKg)
        .distanceToCalories(distanceMeters);
  }

  // Quantidade de passos a partir das calorias gastas
  static int staticCaloriesToSteps(
      double calories, int heightCm, int weightKg) {
    return Calculator(heightCm: heightCm, weightKg: weightKg)
        .caloriesToSteps(calories);
  }

  // Distância (em metros) a partir das calorias gastas
  static double staticCaloriesToDistance(
      double calories, int heightCm, int weightKg) {
    return Calculator(heightCm: heightCm, weightKg: weightKg)
        .caloriesToDistance(calories);
  }
}
